File: android/java/AmpliApp/app/src/androidTest/java/com/example/ampliapp/AmpliTest.java
package com.example.ampliapp;

import android.content.Context;

import androidx.test.platform.app.InstrumentationRegistry;

import org.json.JSONObject;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.junit.MockitoJUnitRunner;

import static org.junit.Assert.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

import com.amplitude.ampli.Ampli;
import com.amplitude.ampli.EventOptions;
import com.amplitude.ampli.EventWithAllPropertiesProperties;
import com.amplitude.ampli.EventWithAllPropertiesRequiredEnum;
import com.amplitude.ampli.IdentifyProperties;
import com.amplitude.ampli.LoadClientOptions;
import com.amplitude.ampli.LoadOptions;
import com.amplitude.api.AmplitudeClient;
import com.amplitude.api.Plan;

/**
 * Instrumented test, which will execute on an Android device.
 *
 * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
 */
@RunWith(MockitoJUnitRunner.class)
public class AmpliTest {
    Ampli ampli;
    Context appContext;

    private final String userId = "test-ampli-user-id";
    private final String deviceId = "test-ampli-device-id";

    @Captor
    ArgumentCaptor<JSONObject> jsonObjectCaptor;

    @Before
    public void setUp() {
        // Context of the app under test.
        this.appContext = InstrumentationRegistry.getInstrumentation().getTargetContext();

        this.ampli = new Ampli();
    }

    @Test
    public void load() {
        AmplitudeClient client = mock(AmplitudeClient.class);

        Plan plan = new Plan().setBranch("branch-1").setSource("source-1").setVersion("version-1");
        this.ampli.load(appContext, new LoadOptions().setClient(new LoadClientOptions().setInstance(client).setPlan(plan)));

        verify(client, times(0)).initialize(any(), any());
        verify(client, times(1)).setPlan(plan);
    }

    @Test
    public void identify() {
        AmplitudeClient client = mock(AmplitudeClient.class);
        this.ampli.load(appContext, new LoadOptions().setClient(new LoadClientOptions().setInstance(client)));

        IdentifyProperties properties = new IdentifyProperties();
        properties.setRequiredNumber(42.0);
        properties.setOptionalArray(new String[]{"A", "ray"});
        this.ampli.identify(
                userId,
                properties,
                new EventOptions().setDeviceId(deviceId).setUserId("some-user")
        );

        verify(client, times(1)).setUserId(userId);
        verify(client, times(1)).setDeviceId(deviceId);

        verify(client, times(1)).setUserProperties(jsonObjectCaptor.capture());
        assertEquals(
                "{\"requiredNumber\":42,\"optionalArray\":[\"A\",\"ray\"]}",
                jsonObjectCaptor.getValue().toString()
        );
    }

    @Test
    public void setGroupSingleValue() {
        AmplitudeClient client = mock(AmplitudeClient.class);
        this.ampli.load(appContext, new LoadOptions().setClient(new LoadClientOptions().setInstance(client)));

        this.ampli.setGroup(
                "group-1",
                "value-1",
                new EventOptions().setDeviceId(deviceId).setUserId(userId)
        );

        verify(client, times(1)).setUserId(userId);
        verify(client, times(1)).setDeviceId(deviceId);
        verify(client, times(1)).setGroup("group-1", "value-1");
    }

    @Test
    public void setGroupMultipleValue() {
        AmplitudeClient client = mock(AmplitudeClient.class);
        this.ampli.load(appContext, new LoadOptions().setClient(new LoadClientOptions().setInstance(client)));

        this.ampli.setGroup(
                "group-1",
                new String[]{"value-1", "value-2", "value-3"},
                new EventOptions().setDeviceId(deviceId).setUserId(userId)
        );

        verify(client, times(1)).setUserId(userId);
        verify(client, times(1)).setDeviceId(deviceId);
        verify(client, times(1)).setGroup("group-1", new String[]{"value-1", "value-2", "value-3"});
    }

    @Test
    public void trackEventNoProperties() {
        AmplitudeClient client = mock(AmplitudeClient.class);
        this.ampli.load(appContext, new LoadOptions().setClient(new LoadClientOptions().setInstance(client)));

        this.ampli.eventNoProperties(
                new EventOptions().setDeviceId(deviceId).setUserId(userId)
        );

        verify(client, times(1)).setUserId(userId);
        verify(client, times(1)).setDeviceId(deviceId);
        verify(client, times(1)).logEvent(eq("Event No Properties"), jsonObjectCaptor.capture());
        assertNull(jsonObjectCaptor.getValue());
    }

    @Test
    public void trackEventWithAllProperties() {
        AmplitudeClient client = mock(AmplitudeClient.class);
        this.ampli.load(appContext, new LoadOptions().setClient(new LoadClientOptions().setInstance(client)));

        EventWithAllPropertiesProperties properties = new EventWithAllPropertiesProperties();
        properties.setRequiredString("Required string");
        properties.setRequiredArray(new String[] { "Required", "array" });
        properties.setRequiredBoolean(true);
        properties.setRequiredEnum(EventWithAllPropertiesRequiredEnum.ENUM2);
        properties.setRequiredInteger(41);
        properties.setRequiredNumber(42);
        this.ampli.eventWithAllProperties(
                properties,
                new EventOptions().setDeviceId(deviceId).setUserId(userId)
        );

        verify(client, times(1)).setUserId(userId);
        verify(client, times(1)).setDeviceId(deviceId);
        verify(client, times(1)).logEvent(
                eq("Event With All Properties"),
                jsonObjectCaptor.capture()
        );
        assertEquals(
                "{\"requiredEnum\":\"Enum2\",\"requiredArray\":[\"Required\",\"array\"],\"requiredNumber\":42,\"requiredConst\":\"some-const-value\",\"requiredBoolean\":true,\"requiredInteger\":41,\"requiredString\":\"Required string\"}",
                jsonObjectCaptor.getValue().toString()
        );
    }

    @Test
    public void flush() {
        AmplitudeClient client = mock(AmplitudeClient.class);
        this.ampli.load(appContext, new LoadOptions().setClient(new LoadClientOptions().setInstance(client)));

        this.ampli.flush();

        verify(client, times(1)).uploadEvents();
    }
}

File: android/kotlin/AmpliApp/app/src/androidTest/java/com/example/ampliapp/AmpliTest.kt
package com.example.ampliapp

import android.content.Context
import androidx.test.platform.app.InstrumentationRegistry
import com.amplitude.ampli.*
import com.amplitude.api.AmplitudeClient
import com.amplitude.api.MiddlewareExtra
import com.amplitude.api.Plan
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentCaptor
import org.mockito.Captor
import org.mockito.Mockito.*
import org.mockito.junit.MockitoJUnitRunner


@RunWith(MockitoJUnitRunner::class)
class AmpliTest {
    lateinit var ampli: Ampli
    lateinit var appContext: Context

    private val userId = "test-ampli-user-id"
    private val deviceId = "test-ampli-device-id"

    @Captor
    lateinit var jsonObjectCaptor: ArgumentCaptor<JSONObject>
    @Captor
    lateinit var extraCaptor: ArgumentCaptor<MiddlewareExtra>

    @Before
    fun setUp() {
        // Context of the app under test.
        this.appContext = InstrumentationRegistry.getInstrumentation().targetContext

        this.ampli = Ampli()
    }

    @Test
    fun load() {
        val client = mock(AmplitudeClient::class.java)

        val plan = Plan().setBranch("branch-1").setSource("source-1").setVersion("version-1")
        this.ampli.load(appContext, LoadOptions(client = LoadClientOptions(instance = client, plan = plan)))

        verify(client, times(0)).initialize(any(), any())
        verify(client, times(1)).setPlan(plan)
    }

    @Test
    fun identify() {
        val client = mock(AmplitudeClient::class.java)
        this.ampli.load(appContext, LoadOptions(client = LoadClientOptions(instance = client)))

        val extra = MiddlewareExtra(mapOf("abc" to 123, "xyz" to "987"))

        this.ampli.identify(
            userId,
            IdentifyProperties(requiredNumber = 42.0, optionalArray = listOf("A", "ray")),
            EventOptions(deviceId = deviceId, userId = "some-user"),
            extra
        )

        verify(client, times(1)).userId = userId
        verify(client, times(1)).deviceId = deviceId

        verify(client, times(1)).setUserProperties(jsonObjectCaptor.capture(), extraCaptor.capture())
        assertEquals(
            """{
  "optionalArray": [
    "A",
    "ray"
  ],
  "requiredNumber": 42
}""", jsonObjectCaptor.value.toString(2)
        )
        assertEquals("{xyz=987, abc=123}", extraCaptor.value.toString())
    }

    @Test
    fun setGroupSingleValue() {
        val client = mock(AmplitudeClient::class.java)
        this.ampli.load(appContext, LoadOptions(client = LoadClientOptions(instance = client)))

        val extra = MiddlewareExtra(mapOf("abc" to 123, "xyz" to "987"))

        this.ampli.setGroup(
            "group-1",
            "value-1",
            EventOptions(deviceId = deviceId, userId = userId),
            extra
        )

        verify(client, times(1)).userId = userId
        verify(client, times(1)).deviceId = deviceId
        verify(client, times(1)).setGroup(eq("group-1"), eq("value-1"), extraCaptor.capture())
        assertEquals("{xyz=987, abc=123}", extraCaptor.value.toString())
    }

    @Test
    fun setGroupMultipleValues() {
        val client = mock(AmplitudeClient::class.java)
        this.ampli.load(appContext, LoadOptions(client = LoadClientOptions(instance = client)))

        val extra = MiddlewareExtra(mapOf("abc" to 123, "xyz" to "987"))

        this.ampli.setGroup(
            "group-1",
            arrayOf("value-1", "value-2", "value-3"),
            EventOptions(deviceId = deviceId, userId = userId),
            extra
        )

        verify(client, times(1)).userId = userId
        verify(client, times(1)).deviceId = deviceId
        verify(client, times(1)).setGroup(eq("group-1"), eq(arrayOf("value-1", "value-2", "value-3")), extraCaptor.capture())
        assertEquals("{xyz=987, abc=123}", extraCaptor.value.toString())
    }

    @Test
    fun trackEventNoProperties() {
        val client = mock(AmplitudeClient::class.java)
        this.ampli.load(appContext, LoadOptions(client = LoadClientOptions(instance = client)))

        val extra = MiddlewareExtra(mapOf("abc" to 123, "xyz" to "987"))

        this.ampli.eventNoProperties(
            EventOptions(deviceId = deviceId, userId = userId),
            extra
        )

        verify(client, times(1)).userId = userId
        verify(client, times(1)).deviceId = deviceId
        verify(client, times(1)).logEvent(
            eq("Event No Properties"),
            jsonObjectCaptor.capture(),
            extraCaptor.capture()
        )
        assertNull(jsonObjectCaptor.value)
        assertEquals("{xyz=987, abc=123}", extraCaptor.value.toString())
    }

    @Test
    fun trackEventWithAllProperties() {
        val client = mock(AmplitudeClient::class.java)
        this.ampli.load(appContext, LoadOptions(client = LoadClientOptions(instance = client)))

        val extra = MiddlewareExtra(mapOf("abc" to 123, "xyz" to "987"))

        this.ampli.eventWithAllProperties(
            EventWithAllPropertiesProperties(
                requiredString = "Required string",
                requiredArray = listOf("Required", "array"),
                requiredBoolean = true,
                requiredEnum = EventWithAllPropertiesRequiredEnum.Enum2,
                requiredInteger = 41,
                requiredNumber = 42.0
            ),
            EventOptions(deviceId = deviceId, userId = userId),
            extra
        )

        verify(client, times(1)).userId = userId
        verify(client, times(1)).deviceId = deviceId
        verify(client, times(1)).logEvent(
            eq("Event With All Properties"),
            jsonObjectCaptor.capture(),
            extraCaptor.capture()
        )
        assertEquals(
            """{
  "requiredArray": [
    "Required",
    "array"
  ],
  "requiredBoolean": true,
  "requiredConst": "some-const-value",
  "requiredEnum": "Enum2",
  "requiredInteger": 41,
  "requiredNumber": 42,
  "requiredString": "Required string"
}""", jsonObjectCaptor.value.toString(2)
        )
        assertEquals("{xyz=987, abc=123}", extraCaptor.value.toString())
    }

    @Test
    fun flush() {
        val client = mock(AmplitudeClient::class.java)
        this.ampli.load(appContext, LoadOptions(client = LoadClientOptions(instance = client)))

        this.ampli.flush()

        verify(client, times(1)).uploadEvents()
    }
}